### PR TITLE
port convert to windows

### PR DESCRIPTION
**Pull request by lplewa**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -77,21 +85,29 @@ macro(add_c_flag flag)
...
-add_c_flag(-Wpointer-arith)
-add_c_flag(-Wsign-compare)
-add_c_flag(-Wsign-conversion)
-add_c_flag(-Wunused-macros)
-add_c_flag(-Wunreachable-code-return)
-
-# Place each function or data item into its own section. Will be used to strip unneeded symbols.
-add_c_flag(-fdata-sections)
-add_c_flag(-ffunction-sections)
+if(NOT WIN32)
```
Comment (marcinslusarz) [inline CMakeLists.txt]: msvc

File: CMakeLists.txt [hunk 2]
```diff
@@ -34,10 +34,16 @@ project(pmdk-convert C)
 
 set(CMAKE_DISABLE_IN_SOURCE_BUILD ON)
 set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${CMAKE_SOURCE_DIR})
 
 include(FindThreads)
 
-set(MIN_VERSION 1.0 CACHE STRING "First supported version (major.minor)")
+if(NOT WIN32)
```
Comment (marcinslusarz) [inline CMakeLists.txt]: it will be easier to understand with reversed condition

File: CMakeLists.txt [hunk 3]
```diff
@@ -77,21 +85,29 @@ macro(add_c_flag flag)
...
+	add_c_flag(-ffunction-sections)
+
+	add_c_flag(-ggdb DEBUG)
+	add_c_flag(-DDEBUG DEBUG)
+
+	add_c_flag(-ggdb RELWITHDEBINFO)
+	add_c_flag(-fno-omit-frame-pointer RELWITHDEBINFO)
+endif()
 
 check_c_compiler_flag(-Wl,-z,relro LINKER_HAS_RELRO)
```
Comment (marcinslusarz) [inline CMakeLists.txt]: this also belongs to !msvc

File: CMakeLists.txt [hunk 4]
```diff
@@ -105,12 +121,6 @@ if(LINKER_HAS_WARN_COMMON)
 	set(CMAKE_SHARED_LINKER_FLAGS  "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--warn-common")
 endif()
 
-add_c_flag(-ggdb DEBUG)
-add_c_flag(-DDEBUG DEBUG)
-
-add_c_flag(-ggdb RELWITHDEBINFO)
-add_c_flag(-fno-omit-frame-pointer RELWITHDEBINFO)
-
 add_c_flag("-U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2" RELEASE)
```
Comment (marcinslusarz) [inline CMakeLists.txt]: and this too

File: CMakeLists.txt [hunk 5]
```diff
@@ -147,6 +157,12 @@ version_supported(1.4 1.5 BUILD_CONVERT_V4)
 version_supported(1.2 1.4 BUILD_CONVERT_V3)
 version_supported(1.1 1.2 BUILD_CONVERT_V2)
 version_supported(1.0 1.1 BUILD_CONVERT_V1)
+if (NOT ${MIN_VERSION} VERSION_GREATER "1.3")
+	set(BUILD_13 1)
+else()
+	set(BUILD_13 0)
+endif()
+message("1.3: ${BUILD_13}")
```
Comment (marcinslusarz) [inline CMakeLists.txt]: ?

File: CMakeLists.txt [hunk 6]
```diff
@@ -243,15 +259,14 @@ if(BUILD_CONVERT_V4)
 	set(pkg nvml-${NVML14}.tar.gz)
 	if(NOT EXISTS ${CMAKE_SOURCE_DIR}/${pkg})
 		message(STATUS "Downloading NVML ${NVML14}")
-		file(DOWNLOAD https://github.com/pmem/pmdk/archive/${NVML14}.tar.gz ${CMAKE_SOURCE_DIR}/${pkg}
-			EXPECTED_HASH SHA256=df7e658e75d28cd80f6d2ff7b9fc9ae2885d52f8923fdbacecfd46215115fb4c)
+		file(DOWNLOAD http://github.com/pmem/pmdk/archive/b7c4d0ae08f7f535e6f8a2b789c2ad3db0ee0858.tar.gz ${CMAKE_SOURCE_DIR}/${pkg}
```
Comment (marcinslusarz) [inline CMakeLists.txt]: why protocol change?

File: CMakeLists.txt [hunk 7]
```diff
@@ -319,37 +348,63 @@ if(BUILD_CONVERT_V4)
 	target_include_directories(pmemobj_convert_v4_o PRIVATE nvml-1.4/src/libpmemcto)
 	target_include_directories(pmemobj_convert_v4_o PRIVATE nvml-1.4/src/common)
 	target_include_directories(pmemobj_convert_v4_o PRIVATE ${PMEM_INCLUDE_DIRS})
+	if(WIN32)
+		target_include_directories(pmemobj_convert_v4_o PRIVATE nvml-1.4/src/windows/include/)
+		target_include_directories(pmemobj_convert_v4_o PRIVATE nvml-1.4/src/windows/getopt/)
+		target_include_directories(pmemobj_convert_v4_o PRIVATE include/)
+		target_compile_options(pmemobj_convert_v4_o PRIVATE /FI"${CMAKE_SOURCE_DIR}/nvml-1.4/src/windows/include/platform.h")
+	endif()
+	if(WIN32)
```
Comment (marcinslusarz) [inline CMakeLists.txt]: empty line between endif and if please

File: CMakeLists.txt [hunk 8]
```diff
@@ -319,37 +348,63 @@ if(BUILD_CONVERT_V4)
 	target_include_directories(pmemobj_convert_v4_o PRIVATE nvml-1.4/src/libpmemcto)
 	target_include_directories(pmemobj_convert_v4_o PRIVATE nvml-1.4/src/common)
 	target_include_directories(pmemobj_convert_v4_o PRIVATE ${PMEM_INCLUDE_DIRS})
+	if(WIN32)
```
Comment (marcinslusarz) [inline CMakeLists.txt]: empty line before if please

File: CMakeLists.txt [hunk 9]
```diff
@@ -319,37 +348,63 @@ if(BUILD_CONVERT_V4)
...
-	add_library(pmemobj_convert_v4_shared SHARED $<TARGET_OBJECTS:pmemobj_convert_v4_o>)
 	target_link_libraries(pmemobj_convert_v4_shared PRIVATE ${PMEM_LIBRARIES})
 	target_link_libraries(pmemobj_convert_v4_shared PRIVATE ${CMAKE_DL_LIBS})
 	target_link_libraries(pmemobj_convert_v4_shared PRIVATE ${CMAKE_THREAD_LIBS_INIT})
-	target_link_libraries(pmemobj_convert_v4_shared PRIVATE -Wl,--version-script=${CMAKE_CURRENT_SOURCE_DIR}/pmemobj_convert.map)
 	# strip unneeded symbols, requires -fdata-sections -ffunction-sections
 	target_link_libraries(pmemobj_convert_v4_shared PRIVATE -Wl,--gc-sections)
 	set_target_properties(pmemobj_convert_v4_shared PROPERTIES OUTPUT_NAME pmemobj_convert_v4)
-
+	set_target_properties(pmemobj_convert_v4_shared PROPERTIES PREFIX "")
```
Comment (marcinslusarz) [inline CMakeLists.txt]: please add comment explaining this line

File: CMakeLists.txt [hunk 10]
```diff
@@ -319,37 +348,63 @@ if(BUILD_CONVERT_V4)
...
 	install(TARGETS pmemobj_convert_v4_shared
 		DESTINATION ${CMAKE_INSTALL_LIBDIR}/pmdk-convert/)
+	if(WIN32)
```
Comment (marcinslusarz) [inline CMakeLists.txt]: empty line

File: CMakeLists.txt [hunk 11]
```diff
@@ -319,37 +348,63 @@ if(BUILD_CONVERT_V4)
...
+		add_library(pmemobj_14 SHARED ${SOURCES_14} ${CMAKE_CURRENT_BINARY_DIR}/pmemobj_14.def)
+		target_compile_options(pmemobj_14 PRIVATE /FI"${CMAKE_SOURCE_DIR}/nvml-1.4/src/windows/include/platform.h")
+	else()
+		add_library(pmemobj_14 SHARED ${SOURCES_14})
+	endif()
 
-	add_library(pmemobj_14 SHARED ${SOURCES_14})
 	target_compile_definitions(pmemobj_14 PRIVATE SRCVERSION="${NVML14}")
 	target_compile_definitions(pmemobj_14 PRIVATE USE_LIBDL)
+	target_compile_definitions(pmemobj_14 PRIVATE PMDK_UTF8_API)
```
Comment (marcinslusarz) [inline CMakeLists.txt]: if (win32)?

File: CMakeLists.txt [hunk 12]
```diff
@@ -319,37 +348,63 @@ if(BUILD_CONVERT_V4)
...
+	target_compile_definitions(pmemobj_14 PRIVATE USE_LIBDL)
```
Comment (marcinslusarz) [inline CMakeLists.txt]: this is duplicated

File: CMakeLists.txt [hunk 13]
```diff
@@ -358,7 +413,7 @@ if(BUILD_CONVERT_V4)
 	set(pkg nvml-${NVML13}.tar.gz)
 	if(NOT EXISTS ${CMAKE_SOURCE_DIR}/${pkg})
 		message(STATUS "Downloading NVML ${NVML13}")
-		file(DOWNLOAD https://github.com/pmem/pmdk/archive/${NVML13}.tar.gz ${CMAKE_SOURCE_DIR}/${pkg}
+		file(DOWNLOAD x ${CMAKE_SOURCE_DIR}/${pkg}
```
Comment (marcinslusarz) [inline CMakeLists.txt]: X?

File: CMakeLists.txt [hunk 14]
```diff
@@ -703,8 +788,15 @@ if(BUILD_CONVERT_V1)
 endif(BUILD_CONVERT_V1)
 
 ###################################################### everything else
-
+if (WIN32)
+# XXX when 1.5 will be added to convert s/nvml-1.4/nvml-$(MAX_VERSION)/
+add_executable(pmdk-convert pmdk-convert.c nvml-1.4/src/windows/getopt/getopt.c)
```
Comment (marcinslusarz) [inline CMakeLists.txt]: indentation

File: pmdk-convert.c [hunk 15]
```diff
@@ -123,6 +125,131 @@ open_lib(const char *name)
...
+		if (*end == '\\' || *end == '/') {
+			*end = '\0';
+			break;
+		}
+	}
+
+	if (end != path) {
+		return path;
+		/* handle edge cases */
+	} else if (*end == '\\' || *end == '/') {
```
Comment (marcinslusarz) [inline pmdk-convert.c]: s/else//

File: tests/CMakeLists.txt [hunk 16]
```diff
@@ -80,61 +93,68 @@ add_executable(create_12 create.c)
...
+target_include_directories(open_13 PRIVATE ${CMAKE_SOURCE_DIR}/nvml-1.3/src/include)
+target_link_libraries(open_13 PRIVATE ${PMEM_LIBRARIES})
+target_link_libraries(open_13 PRIVATE pmemobj_13)
+target_compile_definitions(open_13 PRIVATE NVML_UTF8_API)
+
 set(VERSIONS "${VERSIONS} 1.3")
 test("1.3_pool_conversion" basic13)
+
+if(WIN32)
+	add_custom_command(TARGET open_13 POST_BUILD
```
Comment (marcinslusarz) [inline tests/CMakeLists.txt]: missing comment

File: tests/CMakeLists.txt [hunk 17]
```diff
@@ -80,61 +93,68 @@ add_executable(create_12 create.c)
...
 
 if(BUILD_CONVERT_V4)
 add_executable(create_14 create.c)
 target_include_directories(create_14 PRIVATE ${CMAKE_SOURCE_DIR}/nvml-1.4/src/include)
 target_link_libraries(create_14 PRIVATE ${PMEM_LIBRARIES})
 target_link_libraries(create_14 PRIVATE pmemobj_14)
-set(VERSIONS "${VERSIONS} 1.4")
-test("1.4_pool_conversion" basic14)
-endif()
+target_compile_definitions(create_14 PRIVATE NVML_UTF8_API)
```
Comment (marcinslusarz) [inline tests/CMakeLists.txt]: if (win32)?

File: tests/CMakeLists.txt [hunk 18]
```diff
@@ -80,61 +93,68 @@ add_executable(create_12 create.c)
...
-target_link_libraries(open_13 PRIVATE ${PMEM_LIBRARIES})
-target_link_libraries(open_13 PRIVATE pmemobj_13)
-endif()
-
-if(BUILD_CONVERT_V4)
 add_executable(open_14 open.c)
 target_include_directories(open_14 PRIVATE ${CMAKE_SOURCE_DIR}/nvml-1.4/src/include)
 target_link_libraries(open_14 PRIVATE ${PMEM_LIBRARIES})
 target_link_libraries(open_14 PRIVATE pmemobj_14)
+target_compile_definitions(open_14 PRIVATE NVML_UTF8_API)
```
Comment (marcinslusarz) [inline tests/CMakeLists.txt]: .

**Comment by marcinslusarz** [comment]
___
*[appveyor.yml, line 33 at r4](https://reviewable.io/reviews/pmem/pmdk-convert/12#-LNjwfrRABT8xi3_Tu_d:-LNjwfrS5djsZ0PXYGiZ:b-hy47ic) ([raw file](https://github.com/pmem/pmdk-convert/blob/84860ce1d3f2d3b0028213902bf194360a15f3df/appveyor.yml#L33)):*
> ```YAML
> - cd build
> - ctest -C %CONFIGURATION% --output-on-failure
> - msbuild INSTALL.vcxproj
> ```

EOF=EOL please
